Upstream slice functions

nmanthey opened the pull request:
This change allows to slice based on functions remaining reachable. This slicing allows to reduce the search space to a user chosen subset of the prorgam.

kroening commented:
(I'll get someone to produce a set of tests)

nmanthey commented:
I can certainly add a test case here, I just learned from https://github.com/diffblue/cbmc/pull/2748 that I can show the goto-program to compare it in the test case.

xbauch commented:
3 small fixes:
1) option parser ignored colon-separated list
2) running slicer multiple times broke an invariant about end-function nodes
3) on test was called with wrong options

smowton commented on regression/goto-instrument/fp-reachability-slice1/test.desc:
```diff
@@ -0,0 +1,8 @@
+--fp-reachability-slice c
+^EXIT=0$
+^SIGNAL=0$
+--
```
This is a bit easy to pass at the moment, especially if the precise pretty-printing of "dead f" were to change. Suggest adding some positive requirements too, to make the test more robust.

smowton commented on regression/goto-instrument/fp-reachability-slice2/test.desc:
```diff
@@ -0,0 +1,10 @@
+--fp-reachability-slice a,c
+^EXIT=0$
+^SIGNAL=0$
+dead d;
```
(for comparison, this is better)

smowton commented on regression/goto-instrument/fp-reachability-slice3/test.desc:
```diff
@@ -0,0 +1,9 @@
+main.c
+--fp-reachability-slice b,c
+^EXIT=0$
+^SIGNAL=0$
```
No meaningful positive conditions

smowton commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -86,13 +89,12 @@ void reachability_slicert::fixedpoint_to_assertions(
-        stack.emplace_back(edge.first, true);
-        stack.emplace_back(
-          cfg.entry_map[std::prev(node.PC)], caller_is_known);
+        if(std::prev(node.PC)->is_function_call())
```
Comment about when this might be false?

xbauch replied:
It really shouldn't happen, but something broke in the slicer during multiple calls and the cfg-predecessor of a successor-of-callsite was not a function call. Now that a new slicer is created it does not happen, so I replaced it back with the original invariant.

smowton commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -193,11 +195,12 @@ void reachability_slicert::slice(goto_functionst &goto_functions)
           !e.reaches_assertion && !e.reachable_from_assertion &&
           !i_it->is_end_function())
           i_it->make_assumption(false_exprt());
+        e.reaches_assertion = e.reachable_from_assertion = false;
```
Comment in-inline about why these are reset here?

smowton replied:
(It looks like the slicer might be run more than once, in which case perhaps it should be reset before use, rather than upon executing the transformation?)

xbauch replied:
It seems that slicer is not really supposed to be called multiple times; or it would require a much more thorough resetting that just the reachability flags. I changed the code to build a new slicer before every call.

smowton commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -240,6 +243,32 @@ void reachability_slicer(
+  goto_modelt &goto_model,
+  const std::string &functions_list)
+{
+  reachability_slicert s;
```
s->slicer?

xbauch replied:
This (and the following) were renamed as suggested. Thanks.

smowton commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -240,6 +243,32 @@ void reachability_slicer(
+{
+  reachability_slicert s;
+
+  std::istringstream multiple_functions(functions_list);
```
multiple_functions->functions_stream?

smowton commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -240,6 +243,32 @@ void reachability_slicer(
+  std::string single_function;
+  while(std::getline(multiple_functions, single_function, ','))
+  {
+    in_function_criteriont p(single_function);
```
p->matching_criterion?

kroening commented:
This is nearly there; it needs a bit of commit squashing, i.e., the tests should come with the code that is tested.

smowton reviewed (approved):
Thanks for the prompt fixes!

Commit:
slicer: add function slicer criterion

This criterion marks all instructions of a given function
as slicer starting point. This is useful when the slicing does
not aim at properties, but path that hit a certain function.

Commit:
slicer: add function path slicer

This slicer keeps any function in the program, in case it is backward
or forward reachable from the target function. The analysis can be
repeated for multiple functions.

Note: This implementation cannot deal with order of function calls, e.g.
if functions a,b and c should be kept, then a path that has
a, which calls c, which calls b
will also be kept. However, path where not all functions will be called,
will be dropped.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: ed85c7f55bf7a03e50704c0b6afa90665ffd5b5a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86137585
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

peterschrammel commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1610,6 +1619,9 @@ void goto_instrument_parse_optionst::help()
     " --render-cluster-function    clusterises the dot by functions\n"
     "\n"
     "Slicing:\n"
+    " --fp-reachability-slice <f>  remove instructions that cannot appear on a\n" // NOLINT(*)
```
`--fp-reachability-slice f`
`The list of functions has to be given as a comma-separated list f.`

peterschrammel replied:
Shouldn't this use `HELP_REACHABILITY_SLICER` from below?

xbauch replied:
It seems that this help macro is for CBMC options, e.g. `--reachability-slice-fb` is not a valid option for `goto-instrument`.

peterschrammel commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -240,6 +242,31 @@ void reachability_slicer(
+{
+  std::istringstream functions_stream(functions_list);
+  std::string single_function;
+  while(std::getline(functions_stream, single_function, ','))
```
It would be better to provide the splitting as a `list<string> cmdlinet::get_comma_separated_values(string)` function.

xbauch replied:
done.

peterschrammel commented on src/goto-instrument/reachability_slicer.h:
```diff
@@ -33,13 +37,16 @@ void reachability_slicer(
+
+#define HELP_REACHABILITY_SLICER                                               \
+  " --fp-reachability-slice <f>  remove instructions that cannot appear on a " \
+  "trace that visits all given functions. The list of functions has to be "    \
```
The text help text should be broken after 80 chars and properly aligned (e.g. `trace` should start below `remove`).

xbauch replied:
done.

peterschrammel commented on src/goto-instrument/reachability_slicer.cpp:
```diff
@@ -240,6 +242,31 @@ void reachability_slicer(
+/// Perform reachability slicing on goto_model for selected functions.
+/// \param goto_model Goto program to slice
+/// \param functions_list The functions relevant for the slicing (i.e. starting
+/// point for the search in the cfg). Anything that is reachable in the CFG
```
CFG

xbauch replied:
done.

peterschrammel reviewed (commented):
A couple of nitpicks

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: bfe3dd2fb28d524e936f038c17d5e58f51ca9ef3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86266130
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

peterschrammel commented on src/goto-instrument/reachability_slicer.h:
```diff
@@ -33,13 +37,20 @@ void reachability_slicer(
+
+#define HELP_REACHABILITY_SLICER                                               \
+  " --fp-reachability-slice f    remove instructions that cannot appear on a " \
+  "trace\n                              that visits all given functions. The " \
```
The layout looks a bit odd. Since clang-format is already off for this section, I'm sure you can make this nicer manually.

Commit:
slicer: add function path to goto-instrument

This commit adds the command line argument handling for the
function path slicer.

To keep all possible exeuction traces that hit both function a and b, an
example invocation could look as follows:

goto-instrument --fp-reachability-slice a,b input output

Note, any trace that touches only a, but not b, will not be kept in the
program.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d86701cf6ebc63a73a59f8cc33057d825f4fe57f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86571560
